README.md

```diff
@@ -37,6 +37,142 @@ Third-party apps should link to the same single library `implementation(project(
+     * you can try to register again directly.
+     */
+    override suspend fun onRegistrationFailed(reason: FailedReason, instance: String) {
+      // TODO: Inform the user that the registration failed and therefore notifications
+      // are not available
+    }
+
+    /**
+     * This application is unregistered by the distributor from receiving push messages
+     */
+    override suspend fun onUnregistered(instance: String) {
+        // TODO: Send an unregister action to your web push server, removing the endpoint
+        updateRegistrationState(false)
+    }
+
+    /**
+     * Update the UI
+     */
+    private suspend fun updateRegistrationState(registered: Boolean) {
+        _events.emit(NewRegistrationState(registered))
+    }
+
+    companion object {
+        private const val TAG = "PushServiceImpl"
+        private val _events = MutableSharedFlow<NewRegistrationState>()
+        val events = _events.asSharedFlow()
+    }
+}
+```
+
+Add your `PushService` implementation to your manifest:
+```xml
+<service android:name="global.covesa.sdk.client.push.PushServiceImpl"
+    android:exported="true">
+    <intent-filter>
+        <action android:name="global.covesa.sdk.PUSH_EVENT"/>
+        <action android:name="org.unifiedpush.android.connector.RAISE_TO_FOREGROUND"/>
+    </intent-filter>
+</service>
+```
+
+### Working with PushManager
+The PushManager, as the name says, manages your app interactions with the push distributor. If you skip this step, your PushServiceImpl will never work.
+
+#### Register push service
+In order to register you need to get the distributor first, so your app knows who distributes the push you receive/send.
+To do that user the `tryUseCurrentOrDefaultDistributor` method from the `PushManager`:
+```kotlin
+PushManager.tryUseCurrentOrDefaultDistributor(activityContext) { success ->
+    // TODO: Next step
+}
+```
+With this method the SDK tries to use the saved distributor else, use the default distributor opening the deeplink "unifiedpush://link"
+It can be used on application startup to register to the distributor.
+If you had already registered to a distributor, this ensure the connection is working.
+If the previous distributor has been uninstalled, it will fallback to the user's default.
+If you register for the first time, it will use the user's default Distributor or the OS will ask what it should use.
+When a distributor is picked, you can then register:
+
+```kotlin
+PushManager.tryUseCurrentOrDefaultDistributor(activityContext) { success ->
+  if (success) {
+    val vapidPubKey = <Your vapid pub key>
+    try {
+        PushManager.register(
+            activity,
+            vapid = vapidPubKey
+        )
+        Log.w(TAG, "UnifiedPush registered with vapid $vapidPubKey")
+    } catch (e: PushManager.VapidNotValidException) {
+        Log.w(TAG, "UnifiedPush failed to register with vapid $vapidPubKey. With exception $e")
+    }
+  }
+}
+```
+
+#### Get systems available distributors
```

Comment:
Maybe we can add a comment that these methods can be used to let the user used a non-default distributor.

A UI have to be implemented to pick one of the distrib listed by getDistributors, then, once selected it is saved with saveDistributor, then it is possible to call register

Reply:
@p1gp1g I agree, this would help the developers. Should I develop this feature?

Reply:
Sure, I think _Get systems available distributors_, _Save new distributor_ and _Get registered distributor_ can be under the same section, with an introduction, like _Use non-default distributor_ or something like this. It can be inspired from the [UnifiedPush documentation](https://unifiedpush.org/kdoc/connector/)